### ISSUE TITLE
Inconsistency in multi_match query

### ISSUE THREAD

**Issue by bittusarkar** [elastic/elasticsearch#16098]
There is a discrepancy in the behaviour of `multi_match` query when it comes to resolving field names containing wildcard characters.

The following query will return zero results if the field `subject` does not exists.

```
{
  "multi_match" : {
    "query":    "this is a test", 
    "fields": [ "subject" ] 
  }
}
```

But the following query will fail with an exception saying `"No fields specified for multi_match query"` if there is no field whose name starts with "ms" in the index being queried. Ideally it should return zero results too instead of failing right?

```
{
  "multi_match" : {
    "query":    "this is a test", 
    "fields": [ "ms*" ] 
  }
}
```

I have a use case where fields are dynamically added to the mapping and users can provide the field name to match search text on. Fields added to the mapping also have a metadata string appended to their names which the user may not be aware about. Hence I need to use wildcards in the field name in a `multi_match` query and there are chances that no field name matches too.

**Comment by clintongormley** [comment]
I agree.  Related to https://github.com/elastic/elasticsearch/issues/12016

**Comment by michel-kraemer** [comment]
I bumped into this too. Is there a workaround that I can apply until this issue has been solved?

**Comment by gmridul** [comment]
Is this fixed now?
I just pulled the code from master and the second query is returning zero results (I ensured that there is no field starting with 'ms').

**Comment by qwerty4030** [comment]
@clintongormley Confirmed this is **not** reproducible on latest master with the following:
```
POST test-index/test-type/test-doc1
{
  "test-key":"test-value"
}

POST test-index/_search
{
  "query": {
    "multi_match": {
      "query": "test-query",
      "fields": [
        "doesnt-match*"
      ]
    }
  }
}
results in 0 hits
```
However the same query fails on 2.3.2. Which release will the changes from #12016 land in?

**Comment by clintongormley** [comment]
This is fixed in master thanks to a massive refactoring of how searches are parsed - this obviously can't be backported.  If somebody wants to contribute a quick fix for 2.4, then please submit a PR.

**Comment by andrewla** [comment]
One workaround this that I've found is to add a dummy field to the `fields` specification, like `"fields": ["dummydummydummy", "ms*"]`, which will trigger the behavior on missing fields without causing the wildcard misfiring.  Not sure if this makes the query significantly more expensive.

**Comment by michel-kraemer** [comment]
This issue seems to be fixed in 5.0.0. As far as I am concerned it can be closed.

**Comment by clintongormley** [comment]
Fixed in 5.0, closing
